src/typ/environment.rs

```diff
@@ -466,22 +484,112 @@ impl<'a, 'b> Environment<'a, 'b> {
         }
     }
 
-    pub fn convert_unused_to_warnings(&mut self) {
-        for (name, location) in self.unused_private_types.drain() {
-            self.warnings.push(Warning::UnusedType { name, location })
+    /// Inserts an entity at the current scope for usage tracking.
+    pub fn init_usage(&mut self, name: String, kind: EntityKind, location: SrcSpan) {
+        match self
+            .entity_usages
+            .last_mut()
+            .gleam_expect("Attempted to access non-existant entity usages scope")
+            .insert(name.to_string(), (kind, location, 0))
+        {
+            // PrivateTypes can be overwritten by a constructor with the same name
+            Some((EntityKind::PrivateType, _, _)) => {}
+            Some((kind, location, 0)) => {
+                // an entity was overwritten in the top most scope without being used
+                let mut unused = HashMap::with_capacity(1);
+                let _ = unused.insert(name, (kind, location, 0));
+                self.handle_unused(unused);
+            }
+            _ => {}
         }
-        for (name, location) in self.unused_private_mixed_constructors.drain() {
-            self.warnings
-                .push(Warning::UnusedConstructor { name, location })
+    }
+
+    /// Increments an entity's usage in the current or nearest enclosing scope
+    pub fn increment_usage(&mut self, name: &str) {
+        let mut cascade = None;
+        for scope in self.entity_usages.iter_mut().rev() {
+            match scope.get_mut(name) {
+                Some((EntityKind::PrivateTypeConstructor(type_name), _, usages)) => {
+                    // If a type constructor is used, we consider
+                    // its type also used
```

Comment:
Sorry I'm not understanding this. Could you give an example of the code that this handles? Thank you

Reply:
It cascades the usage of a type constructor to a usage of it's type.
Otherwise this code would leave the private type as unused:
```rust
type A {
  B
}
pub fn a() {
  let b = B
  case b {
    B -> 1
  }
}
```

Reply:
Ah yes, thank you